R/Chrome.R

```diff
@@ -2,6 +2,141 @@
+#'     fulfilled promise.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' async_save_as_pdf <- function(url) {
+#'   function(client) {
+#'     Page <- client$Page
+#'
+#'     Page$enable() %...>% {
+#'       Page$navigate(url = url)
+#'       Page$loadEventFired()
+#'     } %...>% {
+#'       Page$printToPDF()
+#'     } %...>% {
+#'       .$data %>%
+#'         jsonlite::base64_dec() %>%
+#'         writeBin(paste0(httr::parse_url(url)$hostname, ".pdf"))
+#'     }
+#'   }
+#' }
+#'
+#' save_as_pdf <- function(...) {
+#'   list(...) %>%
+#'     purrr::map(async_save_as_pdf) %>%
+#'     chrome_execute(.list = .)
+#' }
+#'
+#' save_as_pdf("https://www.r-project.org/", "https://rstudio.com/")
+#' }
+chrome_execute <- function(
+  ..., .list = NULL, timeouts = 30, cleaning_timeout = 30, async = FALSE,
+  bin = Sys.getenv("HEADLESS_CHROME"), debug_port = 9222L, local = FALSE,
+  extra_args = NULL, headless = TRUE, retry_delay = 0.2, max_attempts = 15L
+) {
+  if(is.null(.list)) {
+    funs <- list(...)
+  } else {
+    stopifnot(is.list(.list))
+    funs <- .list
+  }
+
+  # check arguments
+  purrr::walk(funs, check_is_single_param_fun)
+  assert_that(is.numeric(timeouts))
+  assert_that(is.number(cleaning_timeout))
+
+  # initialize objects
+  timeouts <- rep_len(timeouts, length(funs))
+  total_timeout <- sum(timeouts) + cleaning_timeout
+  results <- vector("list", length(funs))
+  curr_env <- rlang::current_env()
+  execute_fun <- function(client, index, env) {
+    fun <- purrr::pluck(funs, index)
+    delay <- purrr::pluck(timeouts, index)
+    result_saved <- promises::then(fun(client), onFulfilled = function(value) {
+      if(!is.null(value)) {
+        results <- rlang::env_get(env = env, nm = "results")
+        rlang::env_bind(env, results = purrr::assign_in(results, index, value))
+      }
+      client
+    })
+    timeout(result_saved, delay = delay)
+  }
+
+  # launch Chrome
+  chrome <- Chrome$new(
+    bin = bin, debug_port = debug_port, local = local, extra_args = extra_args,
+    headless = headless, retry_delay = retry_delay, max_attempts = max_attempts
+  )
+  client <- chrome$connect()
+  all_funs_executed <- promises::promise_reduce(
+    seq_along(funs), execute_fun, env = curr_env, .init = client
+  )
+  results_available <-
+    promises::then(all_funs_executed, onFulfilled = function(value, env = curr_env) {
+      results <- rlang::env_get(env, "results")
+      if(length(results) == 1L) {
+        results <- results[[1]]
+      }
+      results
+    })
+
+  results_after_cleaning <- promises::finally(results_available, onFinally = function() {
+    # it seems that using hold(), i.e. later::run_now() in finally is not a problem
+    # FMPOV, this seems completely weird, but it works well
+    chrome$close(async = FALSE)
+    })
+
+  promises::catch(results_after_cleaning, onRejected = function(err) {
+    warning(err$message, call. = FALSE)
```

Comment:
Is this suppose to transform error in warning ? 
In the tests I got both... 🤔

Reply:
This is an old code when I tried to close Chrome asynchronously. I think we can safely remove this `catch()`.

Reply:
done in https://github.com/RLesur/crrri/pull/62/commits/24280b9e213ff04528e76de34ab0c6c6ad6ae7b1